Add Functionality to use BackupData using Kopia Server (#1857)
kanisterio/kanister: docs/functions.rst

        args:
          name: "test-snapshot-content-content-dfc8fa67-8b11-4fdf-bf94-928589c2eed8"

.. _backupdatausingkopiaserver:

BackupDataUsingKopiaServer
--------------------------

This function backs up data from a container into any object store
supported by Kanister using Kopia Repository Server as a data mover.

.. note::
   It is important that the application includes a ``kanister-tools``
   sidecar container. This sidecar is necessary to run the
   tools that back up the volume and store it on the object store.

>    Additionally, in order to use this function, a RepositoryServer CR is required

Reviewer: ```suggestion
   Additionally, in order to use this function, a RepositoryServer CR is required.
```

Reviewer: I think we should clearly specify that this repository server CR is needed while creating the actionset. Because, right now it looks like this repo server is expected somewhere in the function args.

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `namespace`, Yes, `string`, namespace of the container that you want to backup the data of
   `pod`, Yes, `string`, pod name of the container that you want to backup the data of
   `container`, Yes, `string`, name of the kanister sidecar container
   `includePath`, Yes, `string`, path of the data to be backed up

>    `snapshotTags`, No, `string`, custom tags to be provided for the kopia snapshots

Reviewer: ```suggestion
   `snapshotTags`, No, `string`, custom tags to be provided to the kopia snapshots
```
?

        args:
          name: "test-snapshot-content-content-dfc8fa67-8b11-4fdf-bf94-928589c2eed8"

.. _backupdatausingkopiaserver:

BackupDataUsingKopiaServer
--------------------------

This function backs up data from a container into any object store

> supported by Kanister using Kopia Repository Server as a data mover.

Reviewer: ```suggestion
supported by Kanister using Kopia Repository Server as data mover.
```

   It is important that the application includes a ``kanister-tools``
   sidecar container. This sidecar is necessary to run the

>    tools that capture path on a volume and store it on the object store.

Reviewer: I might have missed, did we discuss about this already somewhere? Adding sidecar manually in an application doesn't seem to be a good user experience according to me.
If it has already been discussed we will have to improve this document to also explain how to add the sidecar.

Reviewer 2: We are following a similar workflow as of the existing `BackupData` function (https://docs.kanister.io/functions.html#backupdata). We can track this as a separate issue.

Reviewer: yeah makes sense, I understand. @r4rajat  Let's make sure that we raise another PR that explains how do we do that.

   Additionally, in order to use this function, a RepositoryServer CR is required

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

>    `namespace`, Yes, `string`, namespace in which to execute

Reviewer: `namespace in which to execute`
Maybe we should try to improve this. Users don't know what is being executed where. Even I don't know what is this function to going execute to do what.
If this should be application namespace, let's mention application namespace. Same for the other arguments.

Author: Added !

        args:
          name: "test-snapshot-content-content-dfc8fa67-8b11-4fdf-bf94-928589c2eed8"

.. _backupdatausingkopiaserver:

BackupDataUsingKopiaServer
--------------------------

> This function backs up data from a container into any object store
> supported by Kanister using Kopia Repository Server as Data-mover.

Reviewer 2: ```suggestion
This function backs up data from a container into any object store
supported by Kanister using Kopia Repository Server as a data mover.
```

.. note::
   It is important that the application includes a ``kanister-tools``

>    sidecar container. This sidecar is necessary to run the
>    tools that capture path on a volume and store it on the object store.

Reviewer: ```suggestion
   sidecar container. This sidecar is necessary to run the
   tools that back up the volume and store it on the object store.
```

>    Additionally, in order to use this function, a RepositoryServer CR is required

Reviewer 2: We should point to the section on how to setup the kopia repository server setup.

Author: I think we don't have any documentation for that !, will add once the documentation is added

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

>    `namespace`, Yes, `string`, namespace in which to execute (application/deployment namespace)

Reviewer: `namespace in which to execute`
execute what? do we expect users to know what is being executed.
I think we should instead have 

`Namespace of the application that you want to backup the data of`

>    `pod`, Yes, `string`, pod in which to execute

Reviewer: similarly, pod that you want to backup the data of.

make sure to improve edit these according to you.

Author: Done !

>    `container`, Yes, `string`, container in which to execute

Reviewer: this is going to be the sidecar container right?

Author: Yes @viveksinghggits

   `includePath`, Yes, `string`, path of the data to be backed up
   `userPassphrase`, Yes, `string`, user access credentials for kopia repository server
   `certData`, Yes, `string`, certificate data for kopia repository server

> 

Reviewer 2: missing documentation about `snapshotTags` arg.

Author: Added !

>    `namespace`, Yes, `string`, namespace of the application that you want to backup the data of
>    `pod`, Yes, `string`, pod of the application that you want to backup the data of
>    `container`, Yes, `string`, name of the kanister sidecar container
>    `includePath`, Yes, `string`, path of the data to be backed up
>    `userPassphrase`, Yes, `string`, user access credentials for kopia repository server
>    `certData`, Yes, `string`, certificate data for kopia repository server
>    `snapshotTags`, No, `string`, custom tags to be provided for the kopia snapshots

Reviewer: ```suggestion
   `namespace`, Yes, `string`, namespace of the container that you want to backup the data of
   `pod`, Yes, `string`, pod name of the container that you want to backup the data of
   `container`, Yes, `string`, name of the kanister sidecar container
   `includePath`, Yes, `string`, path of the data to be backed up
   `userPassphrase`, Yes, `string`, user access credentials for kopia repository server
   `certData`, Yes, `string`, certificate data for kopia repository server
   `snapshotTags`, No, `string`, custom tags to be provided for the kopia snapshots
```

Sorry, I think I should have been clear before. Since in the function desc here in the docs, we are saying it's used to backup the data from a container. Let's use that container's reference while describing the arguments.
